### ISSUE TITLE
CI pipeline failed after updating to 7.5

### ISSUE THREAD

**Issue by Kemyke** [google/or-tools#1935]
Our project is a dotnet core 3.1 library which referenced the 7.4 versionof OrTools. The CI pipeline build and test this project on an Ubuntu 18.04. With the 7.4 version the project builds successfully and the tests are executed without any failure.

After updating to 7.5, the build step remain successful but the test step is broken. On Windows both steps run without error. The logs indicate that the system doesn't find the shared object. 
Is there some migration steps to update to 7.5?

```
System.TypeInitializationException: The type initializer for 'Google.OrTools.ConstraintSolver.operations_research_constraint_solverPINVOKE' threw an exception. ---> System.TypeInitializationException: The type initializer for 'SWIGExceptionHelper' threw an exception. ---> System.DllNotFoundException: Unable to load shared library 'google-ortools-native' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libgoogle-ortools-native: cannot open shared object file: No such file or directory
  Stack Trace:
      at Google.OrTools.ConstraintSolver.operations_research_constraint_solverPINVOKE.SWIGExceptionHelper.SWIGRegisterExceptionCallbacks_operations_research_constraint_solver(ExceptionDelegate applicationDelegate, ExceptionDelegate arithmeticDelegate, ExceptionDelegate divideByZeroDelegate, ExceptionDelegate indexOutOfRangeDelegate, ExceptionDelegate invalidCastDelegate, ExceptionDelegate invalidOperationDelegate, ExceptionDelegate ioDelegate, ExceptionDelegate nullReferenceDelegate, ExceptionDelegate outOfMemoryDelegate, ExceptionDelegate overflowDelegate, ExceptionDelegate systemExceptionDelegate)
   at Google.OrTools.ConstraintSolver.operations_research_constraint_solverPINVOKE.SWIGExceptionHelper..cctor()
--- End of inner exception stack trace ---
    at Google.OrTools.ConstraintSolver.operations_research_constraint_solverPINVOKE.SWIGExceptionHelper..ctor()
   at Google.OrTools.ConstraintSolver.operations_research_constraint_solverPINVOKE..cctor()
--- End of inner exception stack trace ---
    at Google.OrTools.ConstraintSolver.operations_research_constraint_solverPINVOKE.new_RoutingIndexManager__SWIG_1(Int32 jarg1, Int32 jarg2, Int32 length3, Int32[] jarg3, Int32 length4, Int32[] jarg4)
   at Google.OrTools.ConstraintSolver.RoutingIndexManager..ctor(Int32 num_nodes, Int32 num_vehicles, Int32[] starts, Int32[] ends)
```

If you need more logs or information, please tell me!

**Comment by Mizux** [comment]
Can you try to ldd the google-ortools-native.so to see what is the missing library(ies) ?

We may also need that you provide the RPATH of the lib (e.g. using `objdump -p` IIRC)

**Comment by Mizux** [comment]
well in 7.5 IIRC, we refactor the .net native library management (see #1500) so in order to have one pure .Net package which should be to load any OS dynamic library by usign the trick of **NOT** providing file extension so .Net core will deduce it at runtime, in your case it seems the internal dotnet stuff try to load the lib `libgoogle-ortools-native` without appending the correct extension -> need to investigate this :thinking:

**Comment by Mizux** [comment]
What is the RID of your test project ?

**Comment by Kemyke** [comment]
Meanwhile here is the out of the LDD:

```
******@*********:~/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native# ldd google-ortools-native.so                                                      
        ./google-ortools-native.so: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.22' not found (required by /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libortools.so)
        linux-vdso.so.1 =>  (0x00007ffca99cc000)
        libgflags.so.2.2 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libgflags.so.2.2 (0x00007fed47ddf000)
        libglog.so.0 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libglog.so.0 (0x00007fed47ba3000)
        libprotobuf.so.3.11.2.0 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libprotobuf.so.3.11.2.0 (0x00007fed47294000)
        libCbcSolver.so.3 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libCbcSolver.so.3 (0x00007fed46f99000)
        libCbc.so.3 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libCbc.so.3 (0x00007fed46c6b000)
        libOsiCbc.so.3 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libOsiCbc.so.3 (0x00007fed46a5a000)
        libCgl.so.1 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libCgl.so.1 (0x00007fed46735000)
        libClpSolver.so.1 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libClpSolver.so.1 (0x00007fed464d6000)
        libClp.so.1 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libClp.so.1 (0x00007fed460eb000)
        libOsiClp.so.1 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libOsiClp.so.1 (0x00007fed45e9c000)
        libOsi.so.1 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libOsi.so.1 (0x00007fed45c44000)
        libCoinUtils.so.3 => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libCoinUtils.so.3 (0x00007fed458d9000)
        libortools.so => /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libortools.so (0x00007fed44c21000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fed44a07000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fed447ff000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fed445e2000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fed44260000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fed43f57000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fed43d41000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fed43977000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fed483cc000)
```

**Comment by Kemyke** [comment]
> What is the RID of your test project ?

We don't set the runtime ID manually. The command is `dotnet test` without any `--runtime` option.

**Comment by Kemyke** [comment]
And here is the output of objdump:

```
google-ortools-native.so:     file format elf64-x86-64

Program Header:
    LOAD off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**21
         filesz 0x00000000001a2d0f memsz 0x00000000001a2d0f flags r-x
    LOAD off    0x00000000001a3670 vaddr 0x00000000003a3670 paddr 0x00000000003a3670 align 2**21
         filesz 0x0000000000007768 memsz 0x0000000000007be8 flags rw-
 DYNAMIC off    0x00000000001a5150 vaddr 0x00000000003a5150 paddr 0x00000000003a5150 align 2**3
         filesz 0x0000000000000310 memsz 0x0000000000000310 flags rw-
    NOTE off    0x0000000000000200 vaddr 0x0000000000000200 paddr 0x0000000000000200 align 2**2
         filesz 0x0000000000000024 memsz 0x0000000000000024 flags r--
     TLS off    0x00000000001a3670 vaddr 0x00000000003a3670 paddr 0x00000000003a3670 align 2**3
         filesz 0x0000000000000004 memsz 0x0000000000000020 flags r--
EH_FRAME off    0x0000000000161f60 vaddr 0x0000000000161f60 paddr 0x0000000000161f60 align 2**2
         filesz 0x000000000000b12c memsz 0x000000000000b12c flags r--
   STACK off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**4
         filesz 0x0000000000000000 memsz 0x0000000000000000 flags rw-
   RELRO off    0x00000000001a3670 vaddr 0x00000000003a3670 paddr 0x00000000003a3670 align 2**0
         filesz 0x0000000000002990 memsz 0x0000000000002990 flags r--

Dynamic Section:
  NEEDED               libgflags.so.2.2
  NEEDED               libglog.so.0
  NEEDED               libprotobuf.so.3.11.2.0
  NEEDED               libCbcSolver.so.3
  NEEDED               libCbc.so.3
  NEEDED               libOsiCbc.so.3
  NEEDED               libCgl.so.1
  NEEDED               libClpSolver.so.1
  NEEDED               libClp.so.1
  NEEDED               libOsiClp.so.1
  NEEDED               libOsi.so.1
  NEEDED               libCoinUtils.so.3
  NEEDED               libortools.so
  NEEDED               libz.so.1
  NEEDED               librt.so.1
  NEEDED               libpthread.so.0
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  NEEDED               ld-linux-x86-64.so.2
  RPATH                $ORIGIN:$ORIGIN/../lib:$ORIGIN/../dependencies/install/lib64:$ORIGIN/../dependencies/install/lib
  INIT                 0x00000000000ad758
  FINI                 0x0000000000154c98
  INIT_ARRAY           0x00000000003a3678
  INIT_ARRAYSZ         0x0000000000000080
  FINI_ARRAY           0x00000000003a36f8
  FINI_ARRAYSZ         0x0000000000000008
  GNU_HASH             0x0000000000000228
  STRTAB               0x000000000002e910
  SYMTAB               0x0000000000009e90
  STRSZ                0x000000000006753d
  SYMENT               0x0000000000000018
  PLTGOT               0x00000000003a6000
  PLTRELSZ             0x000000000000df98
  PLTREL               0x0000000000000007
  JMPREL               0x000000000009f7c0
  RELA                 0x00000000000990a0
  RELASZ               0x0000000000006720
  RELAENT              0x0000000000000018
  VERNEED              0x0000000000098f30
  VERNEEDNUM           0x0000000000000007
  VERSYM               0x0000000000095e4e
  RELACOUNT            0x0000000000000077

Version References:
  required from librt.so.1:
    0x09691a75 0x00 14 GLIBC_2.2.5
  required from ld-linux-x86-64.so.2:
    0x0d696913 0x00 12 GLIBC_2.3
  required from libgcc_s.so.1:
    0x0b792650 0x00 09 GCC_3.0
  required from libpthread.so.0:
    0x09691a75 0x00 08 GLIBC_2.2.5
  required from libm.so.6:
    0x09691a75 0x00 06 GLIBC_2.2.5
  required from libc.so.6:
    0x09691a75 0x00 07 GLIBC_2.2.5
    0x06969194 0x00 05 GLIBC_2.14
  required from libstdc++.so.6:
    0x0297f864 0x00 17 GLIBCXX_3.4.14
    0x0297f868 0x00 16 GLIBCXX_3.4.18
    0x0bafd178 0x00 15 CXXABI_1.3.8
    0x0297f867 0x00 13 GLIBCXX_3.4.17
    0x02297f89 0x00 11 GLIBCXX_3.4.9
    0x0297f865 0x00 10 GLIBCXX_3.4.15
    0x056bafd3 0x00 04 CXXABI_1.3
    0x08922974 0x00 03 GLIBCXX_3.4
    0x0297f871 0x00 02 GLIBCXX_3.4.21
```

**Comment by Mizux** [comment]
`./google-ortools-native.so: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.22' not found (required by /home/******/.nuget/packages/google.ortools.runtime.linux-x64/7.5.7466/runtimes/linux-x64/native/./libortools.so)` this one seems bad IMHO

**Comment by Mizux** [comment]
can you run `objdump -p ./usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX`

in a ubuntu:latest docker I've got
```sh
objdump -p ./usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX
2 0x00 0x08922974 GLIBCXX_3.4
3 0x00 0x02297f81 GLIBCXX_3.4.1
	GLIBCXX_3.4 
4 0x00 0x02297f82 GLIBCXX_3.4.2
	GLIBCXX_3.4.1 
5 0x00 0x02297f83 GLIBCXX_3.4.3
	GLIBCXX_3.4.2 
6 0x00 0x02297f84 GLIBCXX_3.4.4
	GLIBCXX_3.4.3 
7 0x00 0x02297f85 GLIBCXX_3.4.5
	GLIBCXX_3.4.4 
8 0x00 0x02297f86 GLIBCXX_3.4.6
	GLIBCXX_3.4.5 
9 0x00 0x02297f87 GLIBCXX_3.4.7
	GLIBCXX_3.4.6 
10 0x00 0x02297f88 GLIBCXX_3.4.8
	GLIBCXX_3.4.7 
11 0x00 0x02297f89 GLIBCXX_3.4.9
	GLIBCXX_3.4.8 
12 0x00 0x0297f860 GLIBCXX_3.4.10
	GLIBCXX_3.4.9 
13 0x00 0x0297f861 GLIBCXX_3.4.11
	GLIBCXX_3.4.10 
14 0x00 0x0297f862 GLIBCXX_3.4.12
	GLIBCXX_3.4.11 
15 0x00 0x0297f863 GLIBCXX_3.4.13
	GLIBCXX_3.4.12 
16 0x00 0x0297f864 GLIBCXX_3.4.14
	GLIBCXX_3.4.13 
17 0x00 0x0297f865 GLIBCXX_3.4.15
	GLIBCXX_3.4.14 
18 0x00 0x0297f866 GLIBCXX_3.4.16
	GLIBCXX_3.4.15 
19 0x00 0x0297f867 GLIBCXX_3.4.17
	GLIBCXX_3.4.16 
20 0x00 0x0297f868 GLIBCXX_3.4.18
	GLIBCXX_3.4.17 
21 0x00 0x0297f869 GLIBCXX_3.4.19
	GLIBCXX_3.4.18 
22 0x00 0x0297f870 GLIBCXX_3.4.20
	GLIBCXX_3.4.19 
23 0x00 0x0297f871 GLIBCXX_3.4.21
	GLIBCXX_3.4.20 
24 0x00 0x0297f872 GLIBCXX_3.4.22
	GLIBCXX_3.4.21 
25 0x00 0x0297f873 GLIBCXX_3.4.23
	GLIBCXX_3.4.22 
26 0x00 0x0297f874 GLIBCXX_3.4.24
	GLIBCXX_3.4.23 
27 0x00 0x0297f875 GLIBCXX_3.4.25
	GLIBCXX_3.4.24 
```

**Comment by Kemyke** [comment]
Here is the output of objdump

```
*******@*******:~# objdump -p /usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX
2 0x00 0x08922974 GLIBCXX_3.4
3 0x00 0x02297f81 GLIBCXX_3.4.1
        GLIBCXX_3.4
4 0x00 0x02297f82 GLIBCXX_3.4.2
        GLIBCXX_3.4.1
5 0x00 0x02297f83 GLIBCXX_3.4.3
        GLIBCXX_3.4.2
6 0x00 0x02297f84 GLIBCXX_3.4.4
        GLIBCXX_3.4.3
7 0x00 0x02297f85 GLIBCXX_3.4.5
        GLIBCXX_3.4.4
8 0x00 0x02297f86 GLIBCXX_3.4.6
        GLIBCXX_3.4.5
9 0x00 0x02297f87 GLIBCXX_3.4.7
        GLIBCXX_3.4.6
10 0x00 0x02297f88 GLIBCXX_3.4.8
        GLIBCXX_3.4.7
11 0x00 0x02297f89 GLIBCXX_3.4.9
        GLIBCXX_3.4.8
12 0x00 0x0297f860 GLIBCXX_3.4.10
        GLIBCXX_3.4.9
13 0x00 0x0297f861 GLIBCXX_3.4.11
        GLIBCXX_3.4.10
14 0x00 0x0297f862 GLIBCXX_3.4.12
        GLIBCXX_3.4.11
15 0x00 0x0297f863 GLIBCXX_3.4.13
        GLIBCXX_3.4.12
16 0x00 0x0297f864 GLIBCXX_3.4.14
        GLIBCXX_3.4.13
17 0x00 0x0297f865 GLIBCXX_3.4.15
        GLIBCXX_3.4.14
18 0x00 0x0297f866 GLIBCXX_3.4.16
        GLIBCXX_3.4.15
19 0x00 0x0297f867 GLIBCXX_3.4.17
        GLIBCXX_3.4.16
20 0x00 0x0297f868 GLIBCXX_3.4.18
        GLIBCXX_3.4.17
21 0x00 0x0297f869 GLIBCXX_3.4.19
        GLIBCXX_3.4.18
22 0x00 0x0297f870 GLIBCXX_3.4.20
        GLIBCXX_3.4.19
23 0x00 0x0297f871 GLIBCXX_3.4.21
        GLIBCXX_3.4.20
```
But i made a mistanke, our build servers haven't been updated yet, so there is 16.04 Ubuntu on them.

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
```
Is it a supported OS for version 7.5?

**Comment by Mizux** [comment]
IIRC To build the linux-x64 lib of the dotnet package I used a centos docker image
see: https://github.com/google/or-tools/blob/stable/tools/docker/centos-8.Dockerfile

!!! Not Tested !!!!
What you may do, is download the or-tools_ubuntu-16.04_v7.5.7466.tar.gz on the release page
https://github.com/google/or-tools/releases/tag/v7.5
It should contains the dotnet package built against ubuntu 16.04 docker image so you may use this package instead or copy the lib inside it to patch your dotnet package cache.

For next release I'll try to figure out how I can provide a linux-x64 Runtime IDentifier compiled against an "old enough" GLIBCXX to be more "portable" !
Or  I Need to test the package generated against all distro (docker will be my friend) to see if it works on all supported platform.

**Comment by ptr1120** [comment]
> well in 7.5 IIRC, we refactor the .net native library management (see #1500) so in order to have one pure .Net package which should be to load any OS dynamic library by usign the trick of **NOT** providing file extension so .Net core will deduce it at runtime, in your case it seems the internal dotnet stuff try to load the lib `libgoogle-ortools-native` without appending the correct extension -> need to investigate this

@Mizux : the dotnet native library management change has been released in ortools 7.4. Additionally I would not expect many problems using a specific linux runtime as long as the required dotnet core dependencies are correctly installed (we are using debian 8 as build system).

**Comment by GuyBenhaim** [comment]
Hello,
Are you going to solve the issues around missing google-ortools-native.dll in v7.6 when building in .NetFramework and .NetCore  Web projects?   (Not console)
Thx
